lib/client/session_windows.go

```diff
@@ -45,22 +230,585 @@ func newSession(
+	//
+	// Unfortunately at the moment the Go SSH library Teleport uses does not
+	// support sending SSH_MSG_DISCONNECT. Instead we close the SSH channel and
+	// SSH client, and try and exit as gracefully as possible.
+	return ns.regularSession(func(s *ssh.Session) error {
+		var err error
+
+		runContext, cancel := context.WithCancel(context.Background())
+		go func() {
+			defer cancel()
+			err = s.Run(strings.Join(cmd, " "))
+		}()
+
+		select {
+		// Run returned a result, return that back to the caller.
+		case <-runContext.Done():
+			return trace.Wrap(err)
+		// The passed in context timed out. This is often due to the user hitting
+		// Ctrl-C.
+		case <-ctx.Done():
+			err = s.Close()
+			if err != nil {
+				log.Debugf("Unable to close SSH channel: %v", err)
+			}
+			err = ns.NodeClient().Client.Close()
+			if err != nil {
+				log.Debugf("Unable to close SSH client: %v", err)
+			}
+			return trace.ConnectionProblem(ctx.Err(), "connection canceled")
+		}
+	})
+}
+
+// watchSignals register UNIX signal handlers and properly terminates a remote shell session
+// must be called as a goroutine right after a remote shell is created
+func (ns *NodeSession) watchSignals(shell io.Writer) {
+	exitSignals := make(chan os.Signal, 1)
+	// catch SIGTERM
+	signal.Notify(exitSignals, syscall.SIGTERM)
+	go func() {
+		defer ns.closer.Close()
+		<-exitSignals
+	}()
+
+	// Note: On Windows, ReadConsoleInput captures most signal sequences
+	// (including Ctrl-Z and Ctrl-C) and emits them directly as VT sequences
+	// which we can pipe to the remote.
+
+	// Catch the interrupt signals. Note that this signal is never triggered by
+	// Ctrl-C in Windows when terminal input processing is disabled. We'll keep
+	// this handler to respond as expected if signalled externally.
+	ctrlCSignal := make(chan os.Signal, 1)
+	signal.Notify(ctrlCSignal, syscall.SIGINT)
+	go func() {
+		for {
+			<-ctrlCSignal
+			_, err := shell.Write([]byte{3})
+			if err != nil {
+				log.Errorf(err.Error())
+			}
+		}
+	}()
+
+	// Note: Windows does not support SIGTSTP (Ctrl-Z), but our use of
+	// ReadConsoleInput ensures the key combo is sent anyway.
+}
+
+// pipeInOut launches two goroutines: one to pipe the local input into the remote shell,
+// and another to pipe the output of the remote shell into the local output
+func (ns *NodeSession) pipeInOut(shell io.ReadWriteCloser) {
+	// Read raw console events via the Windows API.
+	go tncon.ReadInputContinuous()
+
+	// copy from the remote shell to the local output
+	go func() {
+		defer ns.closer.Close()
+		_, err := io.Copy(ns.stdout, shell)
+		if err != nil {
+			log.Errorf(err.Error())
+		}
+	}()
+
+	pipeRead, pipeWrite := io.Pipe()
+
+	// Subscribe to console events and pass them along to the pipe.
+	go func() {
+		defer ns.closer.Close()
+
+		events := tncon.Subscribe()
+		for event := range events {
+			switch ev := event.(type) {
+			case tncon.SequenceEvent:
```

Comment:
what about other events?
if we don't care about them, maybe it's worth simplifying all the C code to just read sequences?

Reply:
Resize events are useful but it was challenging to staple those into the resize handling I'd already written; I'll take another pass at them as I refactor.

I'd initially added mouse events expecting we'd need to implement that ourselves, but Windows is evidently translating those for us. I'll make sure they aren't needed and remove them (along with focus events).

Reply:
I've now removed mouse and focus events.